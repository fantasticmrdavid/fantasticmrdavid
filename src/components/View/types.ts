export type StateProps = {
  location: string,
  isLoading: boolean,
  isFirstLoad: boolean,
  popup: any,
};

export type DispatchProps = {
  updateLocation: (newLocation: string) => void,
};

export type ContainerProps = {
  location?: string,
  target?: string
};
