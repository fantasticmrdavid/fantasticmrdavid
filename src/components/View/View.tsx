import React, { memo, useState } from "react";
import { LogoMenu } from "components/LogoMenu/LogoMenu";
import { Home } from "views/Home/Home";
import { About } from "views/About/About";
import { Work } from "views/Work/Work";
import { Contact } from "views/Contact/Contact";
import GlobalStyles from "styles/global";
import { ImagePreloader } from "components/ImagePreloader/ImagePreloader";
import { LOCATIONS } from "routers/constants";
import {
  Container,
  Control,
  Media,
  ShutterTop,
  ShutterBottom,
  MediaLoadingMessage,
} from "./view.styles";

export type UiProps = {
  location: (typeof LOCATIONS)[keyof typeof LOCATIONS];
  target?: string;
};

export const View = memo(({ location, target }: UiProps) => {
  const [isLoading, setIsLoading] = useState(true);

  return (
    <>
      <ImagePreloader
        hideCompletedCount
        images={["/assets/images/legoFlinders.jpg"]}
        completedAction={() => setIsLoading(false)}
      />
      <GlobalStyles />
      <Container>
        <Control>
          <LogoMenu location={location} />
        </Control>

        <Media $location={location}>
          <ShutterTop $isFirstMediaLoading={isLoading} />
          <ShutterBottom $isFirstMediaLoading={isLoading} />
          <MediaLoadingMessage $isFirstMediaLoading={isLoading}>
            Hello!
          </MediaLoadingMessage>
          {location === LOCATIONS.home && <Home isLoading={isLoading} />}
          {location === LOCATIONS.about && <About isLoading={isLoading} />}
          {location === LOCATIONS.work && <Work target={target} />}
          {location === LOCATIONS.contact && <Contact />}
        </Media>
      </Container>
    </>
  );
});
