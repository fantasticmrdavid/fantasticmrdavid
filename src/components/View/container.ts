import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import ReactGA from 'react-ga';
import { AppState } from 'reducers';
import { locationUpdated } from './actions';
import Ui, { Props as UiProps } from './ui';

type StateProps = {
  location: string,
  loading: boolean,
  firstLoad: boolean,
  popup: any,
}

type DispatchProps = {
  updateLocation: (newLocation: string) => void,
}

type ContainerProps = {
  location?: string,
}

const mapStateToProps = (state: AppState, ownProps: ContainerProps) => {
  const { loading, popup } = state;
  const { location } = ownProps;
  return {
    location: state.location.current || location,
    loading: !location && loading.media,
    firstLoad: !location && loading.firstLoad,
    popup,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>, ownProps: UiProps) => {
  return {
    updateLocation: (newLocation: string) => {
      const { location } = ownProps;
      const updatedLocation = location !== newLocation ? newLocation : 'home';

      dispatch(locationUpdated(updatedLocation));

      ReactGA.event({
        category: 'Navigation',
        action: 'Location Updated',
        label: updatedLocation,
      });
    },
  };
};

export default connect<StateProps, DispatchProps, ContainerProps, AppState>(
  mapStateToProps,
  mapDispatchToProps,
)(Ui);
