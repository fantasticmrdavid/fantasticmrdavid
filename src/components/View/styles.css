.View {
  display: flex;
  justify-content: center;
  align-items: stretch;
  min-height: 100%;
  background: $c-dark-charcoal;

  @media (max-width: $l-w-small) {
    display: block;
  }
}

.control {
  width: 40%;
  display: flex;
  justify-content: center;
  align-items: stretch;
  background: #232323;
  color: $c-text-primary;
  transition: 0.3s;

  @media (max-width: $l-w-small) {
    width: 100%;
  }

  @media (min-width: $l-w-medium) {
    min-height: 100vh;
  }
}

.media {
  position: relative;
  display: flex;
  width: 60%;

  @media (max-width: $l-w-small) {
    width: 100%;
  }

  &.work {
    height: 100vh;  /* IE */
  }
}

.mediaShutter {
  position: absolute;
  right: 0px;
  z-index: 16;
  width: 100%;
  height: 0%;
  transition: 0.3s;
  background: black;

  .firstMediaLoading & {
    height: 50%;
  }
}

.mediaShutterTop {
  composes: mediaShutter;
  top: 0px;
}

.mediaShutterBottom {
  composes: mediaShutter;
  bottom: 0px;
}

.mediaLoadingMessage {
  font-family: $ff-heading;
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0px;
  right: 0px;
  justify-content: center;
  align-items: center;
  font-size: 2.5rem;
  font-weight: bold;
  color: white;
  z-index: 17;
  display: none;

  .firstMediaLoading & {
    display: flex;
  }
}
