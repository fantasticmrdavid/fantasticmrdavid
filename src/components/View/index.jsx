import React from 'react';
import PropTypes from 'prop-types';
import PopupContainer from 'containers/PopupContainer';
import LogoMenuContainer from 'containers/LogoMenuContainer';
import Home from 'views/Home';
import About from 'views/About';
import Work from 'views/Work';
import Contact from 'views/Contact';
import styles from './styles.css';

const View = ({ location, loading, firstLoad }) => {
  const firstMediaLoading = loading && firstLoad;
  return (
    <div className={`${styles.View} ${firstMediaLoading && styles.firstMediaLoading}`}>
      <PopupContainer />
      <section className={styles.control}>
        <LogoMenuContainer />
      </section>

      <section className={`${styles.media} ${styles[location]}`}>
        <div className={styles.mediaShutterTop} />
        <div className={styles.mediaShutterBottom} />
        <div className={styles.mediaLoadingMessage}>Hello!</div>
        { location === 'home' && <Home /> }
        { location === 'about' && <About loading={loading} /> }
        { location === 'work' && <Work /> }
        { location === 'contact' && <Contact /> }
      </section>
    </div>
  );
};

View.propTypes = {
  location: PropTypes.string.isRequired,
  loading: PropTypes.bool,
  firstLoad: PropTypes.bool,
};

export default View;
