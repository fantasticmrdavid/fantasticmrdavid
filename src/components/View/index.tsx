import React, { Fragment } from 'react';
import PopupContainer from 'containers/PopupContainer';
import LogoMenuContainer from 'containers/LogoMenuContainer';
import Home from 'views/Home';
import About from 'views/About';
import Work from 'views/Work';
import Contact from 'views/Contact';
import GlobalStyles from 'styles/global';
import * as styles from './styles';

export interface Props {
  location: string,
  loading: boolean,
  firstLoad: boolean,
}

export default ({ location, loading, firstLoad }: Props) => {
  const {
    Container,
    Control,
    Media,
    ShutterTop,
    ShutterBottom,
    MediaLoadingMessage,
  } = styles;
  const firstMediaLoading = loading && firstLoad;
  return (
    <Fragment>
      <GlobalStyles />
      <Container>
        <PopupContainer />
        <Control>
          <LogoMenuContainer />
        </Control>

        <Media location={location}>
          <ShutterTop firstMediaLoading={firstMediaLoading} />
          <ShutterBottom firstMediaLoading={firstMediaLoading} />
          <MediaLoadingMessage firstMediaLoading={firstMediaLoading}>Hello!</MediaLoadingMessage>
          { location === 'home' && <Home /> }
          { location === 'about' && <About loading={loading} /> }
          { location === 'work' && <Work /> }
          { location === 'contact' && <Contact /> }
        </Media>
      </Container>
    </Fragment>
  );
};
