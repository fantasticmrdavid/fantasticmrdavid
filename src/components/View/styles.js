import styled from 'styled-components';
import theme from 'styles/theme';
import { media } from 'styles/utils';

export const Container = styled.div`
  display: block;
  justify-content: center;
  align-items: stretch;
  min-height: 100%;
  background-color: ${theme.colorDarkCharcoal};

  ${media.small`
    display: flex;
  `}
`;

export const Control = styled.section`
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: stretch;
  background: ${theme.colorSidebar};
  color: ${theme.colorTextPrimary};
  transition: 0.3s;

  ${media.small`
    width: 40%;
  `};

  ${media.medium`
    min-height: 100vh;
  `};
`;

export const Media = styled.section`
  position: relative;
  display: flex;
  width: 100%;
  height: ${props => (props.location === 'work' ? '100vh' : undefined)};

  ${media.small`
    width: 60%;
  `}
`;

const Shutter = styled.div`
  position: absolute;
  right: 0px;
  z-index: 16;
  width: 100%;
  height: ${props => (props.firstMediaLoading ? '50%' : '0%')};
  transition: 0.3s;
  background: ${theme.colorBlack};
`;

export const ShutterTop = styled(Shutter)`
  top: 0px;
`;

export const ShutterBottom = styled(Shutter)`
  bottom: 0px;
`;

export const MediaLoadingMessage = styled.div`
  display: ${props => (props.firstMediaLoading ? 'flex' : 'none')};
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0px;
  right: 0px;
  justify-content: center;
  align-items: center;
  font-family: ${theme.fontHeading};
  font-size: 2.5rem;
  font-weight: 700;
  color: ${theme.colorWhite};
  z-index: 17;
`;
