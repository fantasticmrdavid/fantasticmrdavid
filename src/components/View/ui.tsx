import React, { Fragment, memo } from 'react';
import LogoMenu from 'components/LogoMenu';
import Popup from 'components/Popup';
import Home from 'views/Home';
import About from 'views/About';
import Work from 'views/Work';
import Contact from 'views/Contact';
import GlobalStyles from 'styles/global';
import * as styles from './styles';

export interface Props {
  location: string,
  loading: boolean,
  firstLoad: boolean,
}

const {
  Container,
  Control,
  Media,
  ShutterTop,
  ShutterBottom,
  MediaLoadingMessage,
} = styles;

export default memo(
  ({ location, loading, firstLoad }: Props) => {
    const firstMediaLoading = loading && firstLoad;
    return (
      <Fragment>
        <GlobalStyles />
        <Container>
          <Popup />
          <Control>
            <LogoMenu location={location} />
          </Control>

          <Media location={location}>
            <ShutterTop firstMediaLoading={firstMediaLoading} />
            <ShutterBottom firstMediaLoading={firstMediaLoading} />
            <MediaLoadingMessage firstMediaLoading={firstMediaLoading}>Hello!</MediaLoadingMessage>
            { location === 'home' && <Home /> }
            { location === 'about' && <About loading={loading} /> }
            { location === 'work' && <Work /> }
            { location === 'contact' && <Contact /> }
          </Media>
        </Container>
      </Fragment>
    );
  },
);
