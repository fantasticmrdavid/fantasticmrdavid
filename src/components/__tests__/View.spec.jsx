import React from 'react';
import { shallow } from 'enzyme';
import View from 'components/View/ui';

const props = {
  location: 'about',
  loading: false,
  firstLoad: false,
};

describe('Component -> View', () => {
  it('should render the View without throwing an error', () => {
    expect(shallow(<View {...props} />)).toMatchSnapshot();
  });

  it('should render the View in loading state without throwing an error', () => {
    expect(shallow(<View {...props} loading />)).toMatchSnapshot();
  });

  it('should render the View in first loading state without throwing an error', () => {
    expect(shallow(<View {...props} firstLoad loading />)).toMatchSnapshot();
  });
});
