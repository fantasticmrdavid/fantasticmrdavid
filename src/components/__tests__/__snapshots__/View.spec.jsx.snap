// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> View should render the View in first loading state without throwing an error 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styles__Container>
    <Connect(Component) />
    <styles__Control>
      <Connect(Component)
        location="about"
      />
    </styles__Control>
    <styles__Media
      location="about"
    >
      <styles__ShutterTop
        firstMediaLoading={true}
      />
      <styles__ShutterBottom
        firstMediaLoading={true}
      />
      <styles__MediaLoadingMessage
        firstMediaLoading={true}
      >
        Hello!
      </styles__MediaLoadingMessage>
      <Memo()
        loading={true}
      />
    </styles__Media>
  </styles__Container>
</Fragment>
`;

exports[`Component -> View should render the View in loading state without throwing an error 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styles__Container>
    <Connect(Component) />
    <styles__Control>
      <Connect(Component)
        location="about"
      />
    </styles__Control>
    <styles__Media
      location="about"
    >
      <styles__ShutterTop
        firstMediaLoading={false}
      />
      <styles__ShutterBottom
        firstMediaLoading={false}
      />
      <styles__MediaLoadingMessage
        firstMediaLoading={false}
      >
        Hello!
      </styles__MediaLoadingMessage>
      <Memo()
        loading={true}
      />
    </styles__Media>
  </styles__Container>
</Fragment>
`;

exports[`Component -> View should render the View without throwing an error 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styles__Container>
    <Connect(Component) />
    <styles__Control>
      <Connect(Component)
        location="about"
      />
    </styles__Control>
    <styles__Media
      location="about"
    >
      <styles__ShutterTop
        firstMediaLoading={false}
      />
      <styles__ShutterBottom
        firstMediaLoading={false}
      />
      <styles__MediaLoadingMessage
        firstMediaLoading={false}
      >
        Hello!
      </styles__MediaLoadingMessage>
      <Memo()
        loading={false}
      />
    </styles__Media>
  </styles__Container>
</Fragment>
`;
